### ISSUE TITLE
FastBlockPlacement and FastRightClick don't work after upgrade

### ISSUE THREAD

**Issue by rhocairn** [maruohon/tweakeroo#485]
FastBlockPlacement and FastRightClick no longer work at all after upgrading to tweakeroo-fabric-1.20.4-0.19.1.

I had been playing on MC 1.20.2 with tweakeroo-fabric-1.20.2-0.18.0. I took a couple months off and while I was away my server upgraded to MC 1.20.4, so I had to upgrade too. I'm now on tweakeroo-fabric-1.20.4-0.19.1. I copied over my configs. FastBlockPlacement and FastRightClick no longer work at all (if I toggle them on, right click operates as in vanilla). PlacementRestriction is turned off. I use MultiMC to manage my minecraft versions and started with the Simply Optimized modpack w/fabric loader 0.15.5.

In trying to troubleshoot on my end, I reset the tweakeroo config to default, disabled all other mods (except malilib), and tried in single player. None of these fixed the issue. Then I installed a fresh vanilla + Fabric (0.15.6) in MultiMC, added only malilib and tweakeroo (same versions as before), created a single player creative world, and still had the issue. As a coherence check, I opened the old install (1.20.2) to make sure it was still working there and it was working there, so it is an issue with the upgrade

OS: MacOS X 11.7.10
MultiMC Version: 0.7.0-3825
Versions that **do not** work after upgrading:
- Minecraft 1.20.4
- tweakeroo-fabric-1.20.4-0.19.1
- malilib-fabric-1.20.4-0.18.0.jar
- Fabric loader 0.15.5 and 0.15.6

Version that **did** work before upgrading:
- Minecraft 1.20.2
- tweakeroo-fabric-1.20.2-0.18.0
- malilib-fabric-1.20.1-0.16.0.jar
- Fabric loader 0.14.21

**Comment by maruohon** [comment]
I've gotten several reports lately of fast right click and fast block placement not working in 1.20.4. I have no idea what's going on, since they still work fine in my 1.20.4 test instance using mod version 0.19.1.

A few things to note:
- Fast Block Placement takes precedence over fast Right Click, if I remember right (so FRC will not work if FBP is on)
- Fast Block Placement only works if Placement Restriction is enabled, and that happens automatically by default, as long as you don't disable the `placementRestrictionTiedToFast` option (in other words disabling that option basically just breaks FBP)
  - This also means that if you enable FBP, and then manually disable Placement Restriction after that, then neither tweak will work, as per the previous points.
- There are a few different mods that currently conflict with all clicking related features in Tweakeroo. But if you already tested with **only** Tweakeroo and malilib, then... it's not that.
- If FBP would just place one block and stop when you click and hold right click, then you'd need to check what other restriction tweaks or white/blacklists you have enabled.

**Comment by rhocairn** [comment]
Yeah I did try every combination of block restriction I could think to check (each mode, turned on/off, etc, checked the white/blacklists). Is it possibly a Mac specific issue? No clue what you test on or what OS the other bug reports were coming from, and I don't have a linux or windows box to check.

**Comment by maruohon** [comment]
No the OS shouldn't have anything to do with this.

I've gotten one report recently that the BadOptimizations mod has a conflict, and apparently somehow the conflict is "so persistent" that after removing BadOptimizaions you even need to re-install Tweakeroo entirely. To me this makes no sense, unlesss that mod actually modifies other mod jars somehow. I'm skeptical of this, but if you have/had that mod, I guess give it a try?

**Comment by rhocairn** [comment]
I do have that mod installed on my main minecraft install (it's a part of the Simply Optomized modpack). But I also tried a fresh install of MC (installed via MultiMC) without any other mods besides malilib and tweakeroo, so bad optimizations never got introduced to that test setup.

**Comment by sakura-ryoko** [comment]
I just have some questions regarding the "working" configuration.  I was doing a ton of research into some of the code, and noticed that a bunch of of the Input Handler's (Related to Mouse Clicking also) have changed since **malilib-1.20.1-0.16.0** -- which I'm only half surprised that you are using an older version of the mod under 1.20.2 instead of the latest 1.20.1 version.

So, my question is, under the 1.20.2 scenario, does the **malilib-1.20.1-0.16.3** version still work ?

**Comment by Daniel99j2** [comment]
Can confirm BadOptimizations does break mod IF there are other unknown mods....
In a test instance they worked together but in my main it failed with badoptimisations
Mods in main instance:
[mods.txt](https://github.com/maruohon/tweakeroo/files/14296400/mods.txt)

**Comment by sakura-ryoko** [comment]
Just an update, I have just tried 1.20.4 on a Potato 2012 MBP and Fast Block Placement works as intended.  Even with using the "two finger" method for the "Use" function (And I *highly* recommend using a mouse with a right mouse button and wheel)
This MBP is running Catalina 10.15.7.
MultiMC 5 Version 0.7.0-3825
Oracle Java JDK 17.0.10 for Mac (Non ARM64)
MC: 1.20.4
Fabric Loader: 15.6 // API: 0.96.1+1.20.4
Malilib: 0.18.1
Tweakeroo: 0.19.**3** (Which is only a revision I did quick that adds extra debug logging for the fastBlockPlacement workflow, and fixes the Explosion Mixin, and forces the gammaOverride off)

I have another MBP laying around somewhere that is less of a potato, and M1 powered to try, and perhaps I can also try Tweakroo 0.19.2 but my theory is the results might be the same.

**Comment by rhocairn** [comment]
@sakura-ryoko I just tried my older minecraft installs and they do still work with the newer malilib version you recommended (0.16.3). I also saw that a few days ago there was a 1.20.4 update of malilib (0.18.1) so I tried that and it still didn't work. Oh and I do play with an actual mouse, I tried the trackpad and I found it unbearable.

I have been using OpenJDK (openjdk version "20.0.1" 2023-04-18 installed via homebrew). I just tried removing that and using Oracle JDK (java version "17.0.10" 2024-01-16 LTS). Unfortunately the same problem persisted.

Anything else I can try that might be helpful? Should I update to Tweakeroo 0.19.3 and upload logs, or try older other minecraft/library version combinations?

**Comment by sakura-ryoko** [comment]
What is your Fabric API version ?  I'm also starting to think there might be something else at work with your mouse / trackpad controls ...  IDK tbh.  You can try my "logger" version for 1.20.4 ... you'd have to turn on the Keybind Logger from Malilib (A+C menu, setting doesn't get saved) and Tweakeroo's "debug logging" features that I put into the custom build.  I just uploaded it to my GITHUB, and I'm ONLY doing this for your Issue.  Use the default settings for malilib and tweakeroo, and ONLY enable the debug logger's and fastBlockPlacement.  (This link will be removed once you confirm that you have a copy).

https://github.com/sakura-ryoko/tweakeroo/releases/tag/1.20.4-0.19.3-DEBUGLOG

The debugLog setting is under "General" .  This way you can share the logs in this thread of what they are doing when the game attempts to use fastBlockPlacement.  Also, do you happen to have the GL"whatever" setting for your video turned on or off under your MultiMC ?  The video frame buffer besides the "right click" issue is really the main differences between Mac and PC versions of Minecraft from what I can tell.  It doesn't seem to make a lot of sense, but it seems that Minecraft handles "keybind" actions in between video frames, or at least when it's "screenHandler" is set to NULL momentarily; so it can be related to your video settings.

Obviously we can't fix something that's not generally reproducible on our end, so that makes this problem more difficult to troubleshoot, and why I went out of my way to setup Minecraft on some Mac's I have around.  I can only try Catalina or Ventura, though; and 1.20.4 worked for me on both machines.

**Comment by rhocairn** [comment]
@sakura-ryoko Sorry for dropping off, I've been sick and wasn't able to come back to Minecraft until now.  On the up side, I updated my all of my installed mods (same MC version 1.20.4) and it works now so evidently one of the mods fixed the bug (which I presume was in vanilla MC or in fabric, since previously it failed to work even with no additional mods installed).
